reactor3/src/test/java/com/linecorp/armeria/common/reactor3/RequestContextPropagationFluxTest.java

```diff
@@ -0,0 +1,635 @@
+                await().pollDelay(Duration.ofMillis(200)).until(() -> !disposable1.isDisposed());
+                final Disposable disposable2 = flux.subscribe();
+                await().untilAsserted(() -> {
+                    assertThat(disposable1.isDisposed()).isTrue();
+                    assertThat(disposable2.isDisposed()).isTrue();
+                });
+            }
+        } else {
+            final Disposable disposable1 = flux.subscribe();
+            await().pollDelay(Duration.ofMillis(200)).until(() -> !disposable1.isDisposed());
+            final Disposable disposable2 = flux.subscribe();
+            await().untilAsserted(() -> {
+                assertThat(disposable1.isDisposed()).isTrue();
+                assertThat(disposable2.isDisposed()).isTrue();
+            });
+        }
+        assertThat(ctxExists(ctx)).isFalse();
+    }
+
+    @Test
+    void subscriberContextIsNotMissing() {
+        final ClientRequestContext ctx = newContext();
+        final Flux<String> flux;
+
+        flux = Flux.deferContextual(reactorCtx -> {
+            assertThat((String) reactorCtx.get("foo")).isEqualTo("bar");
+            return Flux.just("baz");
+        });
+
+        final Flux<String> flux1 = flux.contextWrite(reactorCtx -> reactorCtx.put("foo", "bar"));
+        StepVerifier.create(flux1)
+                    .expectNextMatches("baz"::equals)
+                    .verifyComplete();
+        assertThat(ctxExists(ctx)).isFalse();
+    }
+
+    @Test
+    void ctxShouldBeCleanUpEvenIfErrorOccursDuringReactorOperationOnSchedulerThread()
+            throws InterruptedException {
+        // Given
+        final ClientRequestContext ctx = newContext();
+        final Flux<String> flux;
+        final Scheduler single = Schedulers.single();
+
+        // When
+        flux = Flux.just("Hello", "Hi")
+                   .subscribeOn(single)
+                   .delayElements(Duration.ofMillis(1000))
+                   .map(s -> {
+                       if ("Hello".equals(s)) {
+                           throw new RuntimeException();
+                       }
+                       return s;
+                   })
+                   .contextWrite(Context.of(RequestContext.class, ctx));
+
+        // Then
+        StepVerifier.create(flux)
+                    .expectError(RuntimeException.class)
+                    .verify();
+
+        final CountDownLatch latch = new CountDownLatch(1);
+        single.schedule(() -> {
+            assertThat(ctxExists(ctx)).isFalse();
+            latch.countDown();
+        });
+        latch.await();
+
+        assertThat(ctxExists(ctx)).isFalse();
+    }
+
+    private static <T> Flux<T> addCallbacks(Flux<T> flux0,
+                                            ClientRequestContext ctx,
+                                            boolean useContextCapture) {
+        // doOnCancel and doFinally do not have context because we cannot add a hook to the cancel.
```

Comment:
nit; I believe this isn't true anymore

```suggestion
```

We can also add the following callbacks for validation
```java
.doOnCancel(() -> assertThat(ctxExists(ctx)).isTrue())
.doFinally(t -> assertThat(ctxExists(ctx)).isTrue())
```